File: src/lib/schema.ts
import { z } from "zod";

export const HEBREW_STEP_NAMES = [
  "מימרא",
  "קשיא",
  "תירוץ",
  "ראיה",
  "דחיה",
  "שאלה",
  "תשובה",
  "מסקנא",
] as const;

export type HebrewStepName = (typeof HEBREW_STEP_NAMES)[number];

export const STEP_ENGLISH: Record<HebrewStepName, string> = {
  מימרא: "Statement",
  קשיא: "Challenge",
  תירוץ: "Resolution",
  ראיה: "Proof",
  דחיה: "Rejection",
  שאלה: "Question",
  תשובה: "Response",
  מסקנא: "Conclusion",
};

export const MACRO_PHASES = [
  "mishnah",
  "opening_question",
  "source_derivation",
  "objection_defense_cycle",
  "alternate_branch",
  "clarification_of_terms",
  "narrative_aggadah",
  "conclusion",
] as const;

export const BRANCH_ROLES = [
  "continues_current_line",
  "opens_new_branch",
  "returns_to_previous_branch",
  "alternative_approach",
  "conclusion_of_branch",
] as const;

export const SCOPE_OF_STEP = [
  "wording",
  "case_definition",
  "legal_rule",
  "proof_mechanism",
  "source_reading",
  "inference",
  "exception_case",
  "general_logic",
] as const;

export const PhraseSchema = z.object({
  phraseNumber: z.number().int(),
  aramaic: z.string(),
  english: z.string(),
  notes: z.string().optional(),
});
export type Phrase = z.infer<typeof PhraseSchema>;

export const KeyTermSchema = z.object({
  term: z.string(),
  meaning: z.string(),
});
export type KeyTerm = z.infer<typeof KeyTermSchema>;

export const MeforeshCommentSchema = z.object({
  source: z.string(),
  ref: z.string(),
  hebrew: z.string(),
  english: z.string().optional(),
  takeaway: z.string(),
});
export type MeforeshComment = z.infer<typeof MeforeshCommentSchema>;

export const MeforshimBlockSchema = z.object({
  rashi: z.array(MeforeshCommentSchema).default([]),
  tosafot: z.array(MeforeshCommentSchema).default([]),
  rishonim: z.array(MeforeshCommentSchema).default([]),
  acharonim: z.array(MeforeshCommentSchema).default([]),
  interplaySummary: z.string().optional(),
});
export type MeforshimBlock = z.infer<typeof MeforshimBlockSchema>;

export const StepSchema = z.object({
  stepNumber: z.number().int(),
  hebrewStepName: z.enum(HEBREW_STEP_NAMES),
  title: z.string(),
  stepSummary: z.string(),
  simpleTranslation: z.string().optional(),
  whatsHappening: z.string(),
  deeperAnalysis: z.string(),
  keyTerms: z.array(KeyTermSchema),
  whatToRemember: z.string().optional(),
  confusionAlert: z.string().optional(),
  whyThisMatters: z.string().optional(),
  phrases: z.array(PhraseSchema).min(1),
  classificationConfidence: z.enum(["High", "Medium", "Low"]).default("High"),
  alternativePossibleLabel: z.string().optional(),
  triggerLanguage: z.string().optional(),
  macroPhase: z.enum(MACRO_PHASES).optional(),
  branchRole: z.enum(BRANCH_ROLES).optional(),
  dependsOnStepNumbers: z.array(z.number().int()).default([]),
  scopeOfStep: z.enum(SCOPE_OF_STEP).optional(),
  relationToPreviousStep: z.string().optional(),
  kashyaTarget: z.string().optional(),
  kashyaAttackLogic: z.string().optional(),
  terutzResolutionType: z.string().optional(),
  terutzHavaAmina: z.string().optional(),
  terutzMaskana: z.string().optional(),
  sheelahInformationSought: z.string().optional(),
  teshuvahAnswerProvided: z.string().optional(),
  raayaObject: z.string().optional(),
  raayaSupportSource: z.string().optional(),
  dechiyaRejectionScope: z.string().optional(),
  dechiyaFlawIdentified: z.string().optional(),
  mimraCoreRuling: z.string().optional(),
  maskanaFinalTakeaway: z.string().optional(),
  meforshim: MeforshimBlockSchema.optional(),
});
export type Step = z.infer<typeof StepSchema>;

export const SugyaBoundarySchema = z.object({
  sugyaNumber: z.number().int(),
  startLine: z.number().int(),
  endLine: z.number().int(),
  topic: z.string(),
  gist: z.string(),
  openingFormula: z.string().optional(),
  firstStepNumber: z.number().int().optional(),
  lastStepNumber: z.number().int().optional(),
});
export type SugyaBoundary = z.infer<typeof SugyaBoundarySchema>;

export const CostBreakdownSchema = z.object({
  totalUSD: z.number(),
  totalInputTokens: z.number(),
  totalOutputTokens: z.number(),
  byPass: z.record(z.number()),
});

export const DafAnalysisSchema = z.object({
  ref: z.string(),
  masechet: z.string(),
  daf: z.number().int(),
  amud: z.enum(["a", "b"]),
  mainTopic: z.string(),
  overview: z.string(),
  sugyaBoundaries: z.array(SugyaBoundarySchema),
  steps: z.array(StepSchema),
  pipelineVersion: z.string(),
  generatedAt: z.string(),
  modelsUsed: z.record(z.string()),
  cost: CostBreakdownSchema.optional(),
});
export type DafAnalysis = z.infer<typeof DafAnalysisSchema>;

export interface DafSourceText {
  ref: string;
  masechet: string;
  daf: number;
  amud: "a" | "b";
  hebrew: string[];
  english: string[];
}

File: py/schema.py
"""Pydantic mirror of v2/src/lib/schema.ts. The JSON output of build.py must match
the TS-side Zod schema byte-for-byte so the frontend renders unchanged."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

HEBREW_STEP_NAMES = (
    "מימרא",
    "קשיא",
    "תירוץ",
    "ראיה",
    "דחיה",
    "שאלה",
    "תשובה",
    "מסקנא",
)

HebrewStepName = Literal[
    "מימרא", "קשיא", "תירוץ", "ראיה", "דחיה", "שאלה", "תשובה", "מסקנא"
]

STEP_ENGLISH: dict[str, str] = {
    "מימרא": "Statement",
    "קשיא": "Challenge",
    "תירוץ": "Resolution",
    "ראיה": "Proof",
    "דחיה": "Rejection",
    "שאלה": "Question",
    "תשובה": "Response",
    "מסקנא": "Conclusion",
}

MacroPhase = Literal[
    "mishnah",
    "opening_question",
    "source_derivation",
    "objection_defense_cycle",
    "alternate_branch",
    "clarification_of_terms",
    "narrative_aggadah",
    "conclusion",
]

BranchRole = Literal[
    "continues_current_line",
    "opens_new_branch",
    "returns_to_previous_branch",
    "alternative_approach",
    "conclusion_of_branch",
]

ScopeOfStep = Literal[
    "wording",
    "case_definition",
    "legal_rule",
    "proof_mechanism",
    "source_reading",
    "inference",
    "exception_case",
    "general_logic",
]

Confidence = Literal["High", "Medium", "Low"]


class Phrase(BaseModel):
    phraseNumber: int
    aramaic: str
    english: str
    notes: str | None = None


class KeyTerm(BaseModel):
    term: str
    meaning: str


class MeforeshComment(BaseModel):
    source: str
    ref: str
    hebrew: str
    english: str | None = None
    takeaway: str
    # Phrase-aligned Hebrew/English for the verbatim meforesh body (optional).
    phrases: list[Phrase] | None = None


class MeforshimBlock(BaseModel):
    rashi: list[MeforeshComment] = Field(default_factory=list)
    tosafot: list[MeforeshComment] = Field(default_factory=list)
    rishonim: list[MeforeshComment] = Field(default_factory=list)
    acharonim: list[MeforeshComment] = Field(default_factory=list)
    interplaySummary: str | None = None


class Step(BaseModel):
    stepNumber: int
    hebrewStepName: HebrewStepName
    title: str
    stepSummary: str
    simpleTranslation: str | None = None
    whatsHappening: str
    deeperAnalysis: str
    keyTerms: list[KeyTerm]
    whatToRemember: str | None = None
    confusionAlert: str | None = None
    whyThisMatters: str | None = None
    phrases: list[Phrase] = Field(default_factory=list)
    classificationConfidence: Confidence = "High"
    alternativePossibleLabel: str | None = None
    triggerLanguage: str | None = None
    macroPhase: MacroPhase | None = None
    branchRole: BranchRole | None = None
    dependsOnStepNumbers: list[int] = Field(default_factory=list)
    scopeOfStep: ScopeOfStep | None = None
    relationToPreviousStep: str | None = None
    kashyaTarget: str | None = None
    kashyaAttackLogic: str | None = None
    terutzResolutionType: str | None = None
    terutzHavaAmina: str | None = None
    terutzMaskana: str | None = None
    sheelahInformationSought: str | None = None
    teshuvahAnswerProvided: str | None = None
    raayaObject: str | None = None
    raayaSupportSource: str | None = None
    dechiyaRejectionScope: str | None = None
    dechiyaFlawIdentified: str | None = None
    mimraCoreRuling: str | None = None
    maskanaFinalTakeaway: str | None = None
    meforshim: MeforshimBlock | None = None


class SugyaBoundary(BaseModel):
    sugyaNumber: int
    startLine: int
    endLine: int
    topic: str
    gist: str
    openingFormula: str | None = None
    firstStepNumber: int | None = None
    lastStepNumber: int | None = None


class CostBreakdown(BaseModel):
    totalUSD: float = 0.0
    totalInputTokens: int = 0
    totalOutputTokens: int = 0
    byPass: dict[str, float] = Field(default_factory=dict)


class DafAnalysis(BaseModel):
    ref: str
    masechet: str
    daf: int
    amud: Literal["a", "b"]
    mainTopic: str
    overview: str
    sugyaBoundaries: list[SugyaBoundary]
    steps: list[Step]
    pipelineVersion: str
    generatedAt: str
    modelsUsed: dict[str, str]
    cost: CostBreakdown | None = None


class DafSourceText(BaseModel):
    ref: str
    masechet: str
    daf: int
    amud: Literal["a", "b"]
    hebrew: list[str]
    english: list[str]
